### PR TITLE
replication use cases

### PR DESCRIPTION
**Pull request by reggie7**
[534](https://github.com/headwirecom/peregrine-cms/issues/534)

### CODE REVIEW

**Comment by GastonGonzalez** [comment]
@reggie7, This is a really good start, but it looks like these use cases mainly deal with happy path use cases. I believe there may be quite a few edge use cases that may exist. For example, do we really want to to remove all pages when a template is removed? Think large site deployments.

Cris added support for versioning. Could we leverage the versioning support to provide a better a replication experience for users? Just a thought, if we consider versioning along with replication, we could likely remove a template, but still allow all pages in the system that use it to still be available since it would be tied to a particular version. In this model, new pages would not be able to use the template, but previously created pages would continue to work. Likewise, template changes that can seriously break a page (i.e. the removal of container) would still work. Old pages could still leverage the template prior to this change and new pages could use the newly updated template.

Give this some thought and double check the uses cases for missing edge cases like this.

Our goal here is really think about how users will use Peregrine's replication system in a production setting. Also, as you re-review these uses cases, you may want to think about some challenges that may come up based on the Local Replication Service (node-to-node) and Local FS Replication Service. 

If you really want to get these use cases fully documented with the edge cases, consider how you would "break" the customer facing site through replication.Try this use both Local Replication Service (node-to-node) and Local FS Replication Service.

**Comment by cmrockwell** [comment]
I'm reviewing this PR wrt issue #534 and a bit confused. This issue talks about setting more definition for replication processes, which is essentially a specification task. A very important and useful task. However, really good specifications or user stories for replication use cases still appear lacking, so what is getting implemented by this PR? 

Looking at the files changed, it is hard to tell
https://github.com/headwirecom/peregrine-cms/pull/562/files 

feature/tenantstate was merged to the develop branches after extensive teamwork to resolve merge conflicts. When I see massive amount of files changed in a PR dealing with feature/tenantstate, I ask, it is happening again? 

I would respectfully suggest applying the changes to a new branch (taken from develop-sling12), and then writing a new PR, so it becomes clearer from the outside what is being done.

**Comment by reggie7** [comment]
Nice catch @cmrockwell - I've pushed an updated version of the PR branch with `develop-sling12` merged into it ~2h ago, but didn't realize the base branch here is still set to `feature/tenantstate`. It's changed now already, so the diff should be clear.

At the moment I'm still working to supply extended descriptions of the replication cases as requested in [the source ticket](https://github.com/headwirecom/peregrine-cms/issues/534) and [the above comment](https://github.com/headwirecom/peregrine-cms/pull/562#issuecomment-698637948). Feel free to look at the 4 **Root Site** cases if you have the time to review this PR - this is what I've recently updated. I'll push other updates soon, hopefully you'll get notification emails about them.

Any suggestions will be welcome as to what edge cases we might encounter within the replication processes, since you've worked in this area, your input will be invaluable.

File: docs/public/replication/asset-deleted.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+## Postconditions:
+1. The Asset is now removed from replication cache.
+1. All unreferenced tags associated with the Asset
+   are now removed from replication cache.
+
+## Main Course:
+1. Author deletes an Asset.
+1. System presents the author with a confirmation dialog asking about removal of
+   replicated resources.
+1. System presents the author a list of other pages that reference the Asset to be removed.
```
Comment (cmrockwell) [inline docs/public/replication/asset-deleted.md]: There may be a gap between what is said and what is meant by this line. Seems to say that a list pages, which use the asset, will be removed. Perhaps what is meant is asset references will be removed from a list of pages.

Comment (GastonGonzalez) [reply docs/public/replication/asset-deleted.md]: @cmrockwell - You are correct. The pages will NOT be removed. The question is really this: 
1. Should we actually remove the references from the various nodes?
2. Or, should we simply inform the user that the list of pages will be impacted if this action is executed?

CC: @reusr1

File: docs/public/replication/index.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+![Replication Strategies](replication-types.jpg)
+
+Repository is quite safe in most cases since the pages can be rendered on the fly.  
+Much more attention has to come to the cache-like replication options, i.e. those, where we store
+final renderings of the replicated resources. A lot can go wrong if we don't handle the references
+properly.
+
+## Versioning Role
+We should version the published resources and use those versions to render published resources that depend on them.
+Let's not though, that it will be much more feasible to simply keep an up-to-date published tree of resources
```
Comment (GastonGonzalez) [inline docs/public/replication/index.md]: Typo: "Let's not" -> "Let's note"

**Comment by GastonGonzalez** [comment]
@reggie7 - Please add a use case for the sitemap. We need to ensure that the sitemap regardless of implementation reflects the current publication state. Thanks. This is top of mind for @reusr1 and will likely be one of the first issues we tackle. Thanks.

**Comment by reggie7** [comment]
Sitemap use cases are restored now, please check @GastonGonzalez - they're quite simplistic since I assume no issues with them - it's basically about writing and removing text files based on existing cache.

**Comment by reggie7** [comment]
Well, we trigger the sitemap update also on resource changes, so a page change does start the update process. There will only be a short time delay coming from debounce interval (think couple of seconds). We just need to create simple hooks that will implement certain interface. That's from code perspective. But I assume we want to have that requirement explicitly added in the use case?

**Pull request merged**